### DISCUSSION TITLE
Numpy speed issue

### DISCUSSION THREAD

**Discussion by ronanRaspberry** [community/community#143104] [Programming Help]
### Select Topic Area

Question

### Body

Hi all,
Recently I started to use Numba ver 0.60.0 with numpy ver: 2.0.0. Since Numba doesn't handle the 'axis' keyword, for example
np.roll(array, 1, axis=0) will fail to compile with numba, I decided to write a replacement that would be acceptable to Numba. (see below) I notice that my replacement is orders of magnitude faster than the out-of-the-box np.roll. ???, I thought numpy was compiled C code, am i missing something ???
rx

# code
import numpy as np
from numba import njit

# replacement for np.roll
@njit
def NumpyRollAx0(arr, roll):
    negRoll = roll * -1   
    if roll >0:
        arr = np.concatenate((arr[negRoll:,:], arr[:negRoll,:]))
    else:
        arr = np.concatenate((arr[negRoll:,:], arr[:negRoll,:]))
    return arr

# test the replacement
arr = np.zeros((4,10))
arr[0] = 0
arr[1] = 1
arr[2] = 2
arr[3] = 3
print('original arr)')
print(arr)
print()
a= np.roll(arr, -2, axis =0)
print('np.roll(arr, -2, axis=0)')
%timeit np.roll(arr, -2, axis =0)
print(a)
b= NumpyRollAx0(arr, -2)

print()
print('NumpyRollAx0(arr, -2)')
%timeit NumpyRollAx0(arr, -2)
print(b)

# output 
original arr)
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]
 [3. 3. 3. 3. 3. 3. 3. 3. 3. 3.]]

np.roll(arr, -2, axis=0)
3.01 µs ± 14.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
[[2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]
 [3. 3. 3. 3. 3. 3. 3. 3. 3. 3.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]

NumpyRollAx0(arr, -2)
433 ns ± 1.19 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
[[2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]
 [3. 3. 3. 3. 3. 3. 3. 3. 3. 3.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]

**Comment by ronanRaspberry** [comment]
Oops, sorry for the lack of indents in the above code, i just copied and pasted from Juyter Notebook and didn't notice that the indents were missing,
rx

**Comment by LiteBrite82** [comment]
Thanks for posting in the GitHub Community, @ronanRaspberry ! 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.
